kernel/src/table_features/mod.rs

```diff
@@ -221,6 +677,276 @@ pub(crate) static SUPPORTED_WRITER_FEATURES: LazyLock<Vec<TableFeature>> = LazyL
+            let Some(info) = feature.info() else {
+                continue;
+            };
+
+            // Only check ReaderWriter features for read support
+            if !matches!(info.feature_type, FeatureType::ReaderWriter) {
+                continue;
+            }
+
+            match &info.read_support {
+                SupportCheck::Supported => {
+                    // Feature is supported for this operation
+                }
+                SupportCheck::NotSupported => {
+                    return Err(Error::unsupported(format!(
+                        "Table feature '{}' is not supported for reading",
+                        feature
+                    )));
+                }
+                SupportCheck::Custom(check_fn) => {
+                    check_fn(self, operation)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Validates that the Rust kernel has write support for the specified data type.
+    ///
+    /// The `operation` parameter specifies what type of data is being written:
+    /// - `Operation::Scan`: Writing regular table data
+    /// - `Operation::Cdf`: Writing to change data feed enabled tables
+    ///
+    /// Returns an error if any supported feature (Writer or ReaderWriter) lacks write support for the given operation.
+    pub(crate) fn validate_write_support(&self, operation: Operation) -> DeltaResult<()> {
+        let supported_features = self.get_supported_features();
+
+        for feature in supported_features {
+            let Some(info) = feature.info() else {
+                continue;
+            };
+
+            // Check both Writer and ReaderWriter features for write support
+            if !matches!(
+                info.feature_type,
+                FeatureType::Writer | FeatureType::ReaderWriter
+            ) {
+                continue;
+            }
+
+            match &info.write_support {
+                SupportCheck::Supported => {
+                    // Feature is supported for this operation
+                }
+                SupportCheck::NotSupported => {
+                    return Err(Error::unsupported(format!(
+                        "Table feature '{}' is not supported for writing",
+                        feature
+                    )));
+                }
+                SupportCheck::Custom(check_fn) => {
+                    check_fn(self, operation)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Validates that all feature dependencies are satisfied.
+    ///
+    /// For each supported feature, checks that:
+    /// 1. All feature requirements are satisfied
+    /// 2. Custom validations pass
+    ///
+    /// Returns an error if any requirement or validation fails.
+    pub(crate) fn validate_dependencies(&self) -> DeltaResult<()> {
+        let supported_features = self.get_supported_features();
+
+        for feature in &supported_features {
```

Comment:
```suggestion
        for feature in self.get_supported_features() {
```

ditto above. there might be some ownership issues, but I think if you change `get_supported_features` as I suggested it'll work, and/or you need an `iter()`